File: Python/OneLang/Generator/TemplateFileGeneratorPlugin.py
from onelang_core import *
from onelang_yaml import *
import OneLang.Parsers.Common.ExpressionParser as exprPars
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.Interfaces as ints
import OneLang.One.Ast.Statements as stats
import OneLang.Generator.IGeneratorPlugin as iGenPlug
import OneLang.Parsers.Common.Reader as read
import OneLang.VM.Values as vals
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Types as types
import OneLang.Template.TemplateParser as templPars
import OneLang.Generator.IGenerator as iGen
import OneLang.VM.ExprVM as exprVM
import OneLang.Parsers.TypeScriptParser as typeScrPars
import OneLang.One.Ast.AstTypes as astTypes

class CodeTemplate:
    def __init__(self, template, includes, if_expr):
        self.template = template
        self.includes = includes
        self.if_expr = if_expr

class CallTemplate:
    def __init__(self, class_name, method_name, args, template):
        self.class_name = class_name
        self.method_name = method_name
        self.args = args
        self.template = template

class FieldAccessTemplate:
    def __init__(self, class_name, field_name, template):
        self.class_name = class_name
        self.field_name = field_name
        self.template = template

class ExpressionValue:
    def __init__(self, value):
        self.value = value
    
    def equals(self, other):
        return isinstance(other, ExpressionValue) and other.value == self.value

class TypeValue:
    def __init__(self, type):
        self.type = type
    
    def equals(self, other):
        return isinstance(other, TypeValue) and astTypes.TypeHelper.equals(other.type, self.type)

class LambdaValue:
    def __init__(self, callback):
        self.callback = callback
    
    def equals(self, other):
        return False
    
    def call(self, args):
        return self.callback(args)

class TemplateFileGeneratorPlugin:
    def __init__(self, generator, template_yaml):
        self.methods = {}
        self.fields = {}
        self.model_globals = {}
        self.generator = generator
        root = OneYaml.load(template_yaml)
        expr_dict = root.dict("expressions")
        if expr_dict == None:
            return
        
        for expr_str in expr_dict.keys():
            val = expr_dict.get(expr_str)
            tmpl_only = val.type() == VALUE_TYPE.STRING
            if_str = None if tmpl_only else val.str("if")
            if_expr = None if if_str == None else typeScrPars.TypeScriptParser2(if_str, None).parse_expression()
            tmpl = CodeTemplate(val.as_str(), [], None) if tmpl_only else CodeTemplate(val.str("template"), val.str_arr("includes"), if_expr)
            
            self.add_expr_template(expr_str, tmpl)
    
    def prop_access(self, obj, prop_name):
        if isinstance(obj, ExpressionValue) and prop_name == "type":
            return TypeValue(obj.value.get_type())
        if isinstance(obj, TypeValue) and prop_name == "name" and isinstance(obj.type, astTypes.ClassType):
            return vals.StringValue(obj.type.decl.name)
        return None
    
    def stringify_value(self, value):
        if isinstance(value, ExpressionValue):
            result = self.generator.expr(value.value)
            return result
        return None
    
    def add_method(self, name, call_tmpl):
        if not (name in self.methods):
            self.methods[name] = []
        # @php $this->methods[$name][] = $callTmpl;
        self.methods.get(name).append(call_tmpl)
    
    def add_expr_template(self, expr_str, tmpl):
        expr = typeScrPars.TypeScriptParser2(expr_str, None).parse_expression()
        if isinstance(expr, exprs.UnresolvedCallExpression) and isinstance(expr.func, exprs.PropertyAccessExpression) and isinstance(expr.func.object, exprs.Identifier):
            call_tmpl = CallTemplate(expr.func.object.text, expr.func.property_name, list(map(lambda x: (x).text, expr.args)), tmpl)
            self.add_method(f'''{call_tmpl.class_name}.{call_tmpl.method_name}@{len(call_tmpl.args)}''', call_tmpl)
        elif isinstance(expr, exprs.UnresolvedCallExpression) and isinstance(expr.func, exprs.Identifier):
            call_tmpl = CallTemplate(None, expr.func.text, list(map(lambda x: (x).text, expr.args)), tmpl)
            self.add_method(f'''{call_tmpl.method_name}@{len(call_tmpl.args)}''', call_tmpl)
        elif isinstance(expr, exprs.PropertyAccessExpression) and isinstance(expr.object, exprs.Identifier):
            field_tmpl = FieldAccessTemplate(expr.object.text, expr.property_name, tmpl)
            self.fields[f'''{field_tmpl.class_name}.{field_tmpl.field_name}'''] = field_tmpl
        elif isinstance(expr, exprs.UnresolvedNewExpression) and isinstance(expr.cls_, astTypes.UnresolvedType):
            call_tmpl = CallTemplate(expr.cls_.type_name, "constructor", list(map(lambda x: (x).text, expr.args)), tmpl)
            self.add_method(f'''{call_tmpl.class_name}.{call_tmpl.method_name}@{len(call_tmpl.args)}''', call_tmpl)
        else:
            raise Error(f'''This expression template format is not supported: \'{expr_str}\'''')
    
    def expr(self, expr):
        is_call_expr = isinstance(expr, exprs.StaticMethodCallExpression) or isinstance(expr, exprs.InstanceMethodCallExpression) or isinstance(expr, exprs.GlobalFunctionCallExpression) or isinstance(expr, exprs.NewExpression)
        is_field_ref = isinstance(expr, refs.StaticFieldReference) or isinstance(expr, refs.StaticPropertyReference) or isinstance(expr, refs.InstanceFieldReference) or isinstance(expr, refs.InstancePropertyReference)
        
        if not is_call_expr and not is_field_ref:
            return None
        # quick return
        
        code_tmpl = None
        model = vals.ObjectValue({})
        context = exprVM.VMContext(model, self)
        
        model.props["type"] = TypeValue(expr.get_type())
        for name in self.model_globals.keys():
            model.props[name] = self.model_globals.get(name)
        
        if is_call_expr:
            call = expr
            parent_intf = call.get_parent_interface()
            method_name = f'''{("" if parent_intf == None else f'{parent_intf.name}.')}{call.get_method_name()}@{len(call.args)}'''
            call_tmpls = self.methods.get(method_name)
            if call_tmpls == None:
                return None
            
            for call_tmpl in call_tmpls:
                if isinstance(expr, exprs.InstanceMethodCallExpression):
                    model.props["this"] = ExpressionValue(expr.object)
                i = 0
                
                while i < len(call_tmpl.args):
                    model.props[call_tmpl.args[i]] = ExpressionValue(call.args[i])
                    i = i + 1
                
                if call_tmpl.template.if_expr == None or (exprVM.ExprVM(context).evaluate(call_tmpl.template.if_expr)).value:
                    code_tmpl = call_tmpl.template
                    break
        elif is_field_ref:
            cm = (expr).get_variable()
            field = self.fields.get(f'''{cm.get_parent_interface().name}.{cm.name}''')
            if field == None:
                return None
            
            if isinstance(expr, refs.InstanceFieldReference) or isinstance(expr, refs.InstancePropertyReference):
                model.props["this"] = ExpressionValue((expr).object)
            code_tmpl = field.template
        else:
            return None
        
        if code_tmpl == None:
            return None
        
        for inc in code_tmpl.includes or []:
            self.generator.add_include(inc)
        
        tmpl = templPars.TemplateParser(code_tmpl.template).parse()
        result = tmpl.format(context)
        return result
    
    def stmt(self, stmt):
        return None

File: PHP/src/OneLang/Generator/TemplateFileGeneratorPlugin.php
<?php

namespace OneLang\Generator\TemplateFileGeneratorPlugin;

use OneLang\Parsers\Common\ExpressionParser\ExpressionParser;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\GlobalFunctionCallExpression;
use OneLang\One\Ast\Expressions\ICallExpression;
use OneLang\One\Ast\Expressions\Identifier;
use OneLang\One\Ast\Expressions\IMethodCallExpression;
use OneLang\One\Ast\Expressions\InstanceMethodCallExpression;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Expressions\StaticMethodCallExpression;
use OneLang\One\Ast\Expressions\UnresolvedCallExpression;
use OneLang\One\Ast\Interfaces\IExpression;
use OneLang\One\Ast\Interfaces\IType;
use OneLang\One\Ast\Statements\Statement;
use OneLang\Generator\IGeneratorPlugin\IGeneratorPlugin;
use OneLang\Parsers\Common\Reader\Reader;
use OneLang\VM\Values\BooleanValue;
use OneLang\VM\Values\ICallableValue;
use OneLang\VM\Values\IVMValue;
use OneLang\VM\Values\ObjectValue;
use OneLang\VM\Values\StringValue;
use OneLang\One\Ast\References\IInstanceMemberReference;
use OneLang\One\Ast\References\InstanceFieldReference;
use OneLang\One\Ast\References\InstancePropertyReference;
use OneLang\One\Ast\References\StaticFieldReference;
use OneLang\One\Ast\References\StaticPropertyReference;
use OneLang\One\Ast\References\VariableReference;
use OneLang\One\Ast\Types\IClassMember;
use OneLang\Template\TemplateParser\TemplateParser;
use OneLang\Generator\IGenerator\IGenerator;
use OneLang\VM\ExprVM\ExprVM;
use OneLang\VM\ExprVM\IVMHooks;
use OneLang\VM\ExprVM\VMContext;
use OneLang\Parsers\TypeScriptParser\TypeScriptParser2;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\TypeHelper;

class CodeTemplate {
    public $template;
    public $includes;
    public $ifExpr;
    
    function __construct($template, $includes, $ifExpr) {
        $this->template = $template;
        $this->includes = $includes;
        $this->ifExpr = $ifExpr;
    }
}

class CallTemplate {
    public $className;
    public $methodName;
    public $args;
    public $template;
    
    function __construct($className, $methodName, $args, $template) {
        $this->className = $className;
        $this->methodName = $methodName;
        $this->args = $args;
        $this->template = $template;
    }
}

class FieldAccessTemplate {
    public $className;
    public $fieldName;
    public $template;
    
    function __construct($className, $fieldName, $template) {
        $this->className = $className;
        $this->fieldName = $fieldName;
        $this->template = $template;
    }
}

class ExpressionValue implements IVMValue {
    public $value;
    
    function __construct($value) {
        $this->value = $value;
    }
    
    function equals($other) {
        return $other instanceof ExpressionValue && $other->value === $this->value;
    }
}

class TypeValue implements IVMValue {
    public $type;
    
    function __construct($type) {
        $this->type = $type;
    }
    
    function equals($other) {
        return $other instanceof TypeValue && TypeHelper::equals($other->type, $this->type);
    }
}

class LambdaValue implements ICallableValue {
    public $callback;
    
    function __construct($callback) {
        $this->callback = $callback;
    }
    
    function equals($other) {
        return false;
    }
    
    function call($args) {
        return $this->callback($args);
    }
}

class TemplateFileGeneratorPlugin implements IGeneratorPlugin, IVMHooks {
    public $methods;
    public $fields;
    public $modelGlobals;
    public $generator;
    
    function __construct($generator, $templateYaml) {
        $this->generator = $generator;
        $this->methods = Array();
        $this->fields = Array();
        $this->modelGlobals = Array();
        $root = OneYaml::load($templateYaml);
        $exprDict = $root->dict("expressions");
        
        foreach (array_keys($exprDict) as $exprStr) {
            $val = @$exprDict[$exprStr] ?? null;
            $ifStr = $val->str("if");
            $ifExpr = $ifStr === null ? null : (new TypeScriptParser2($ifStr, null))->parseExpression();
            $tmpl = $val->type() === ValueType::STRING ? new CodeTemplate($val->asStr(), array(), null) : new CodeTemplate($val->str("template"), $val->strArr("includes"), $ifExpr);
            
            $this->addExprTemplate($exprStr, $tmpl);
        }
    }
    
    function propAccess($obj, $propName) {
        if ($obj instanceof ExpressionValue && $propName === "type")
            return new TypeValue($obj->value->getType());
        if ($obj instanceof TypeValue && $propName === "name" && $obj->type instanceof ClassType)
            return new StringValue($obj->type->decl->name);
        return null;
    }
    
    function stringifyValue($value) {
        if ($value instanceof ExpressionValue) {
            $result = $this->generator->expr($value->value);
            return $result;
        }
        return null;
    }
    
    function addMethod($name, $callTmpl) {
        if (!(array_key_exists($name, $this->methods)))
            $this->methods[$name] = array();
        @$this->methods[$name] ?? null[] = $callTmpl;
    }
    
    function addExprTemplate($exprStr, $tmpl) {
        $expr = (new ExpressionParser(new Reader($exprStr)))->parse();
        if ($expr instanceof UnresolvedCallExpression && $expr->func instanceof PropertyAccessExpression && $expr->func->object instanceof Identifier) {
            $callTmpl = new CallTemplate($expr->func->object->text, $expr->func->propertyName, array_map(function ($x) { return ($x)->text; }, $expr->args), $tmpl);
            $this->addMethod($callTmpl->className . "." . $callTmpl->methodName . "@" . count($callTmpl->args), $callTmpl);
        }
        else if ($expr instanceof UnresolvedCallExpression && $expr->func instanceof Identifier) {
            $callTmpl = new CallTemplate(null, $expr->func->text, array_map(function ($x) { return ($x)->text; }, $expr->args), $tmpl);
            $this->addMethod($callTmpl->methodName . "@" . count($callTmpl->args), $callTmpl);
        }
        else if ($expr instanceof PropertyAccessExpression && $expr->object instanceof Identifier) {
            $fieldTmpl = new FieldAccessTemplate($expr->object->text, $expr->propertyName, $tmpl);
            $this->fields[$fieldTmpl->className . "." . $fieldTmpl->fieldName] = $fieldTmpl;
        }
        else
            throw new \OneLang\Core\Error("This expression template format is not supported: '" . $exprStr . "'");
    }
    
    function expr($expr) {
        $isCallExpr = $expr instanceof StaticMethodCallExpression || $expr instanceof InstanceMethodCallExpression || $expr instanceof GlobalFunctionCallExpression;
        $isFieldRef = $expr instanceof StaticFieldReference || $expr instanceof StaticPropertyReference || $expr instanceof InstanceFieldReference || $expr instanceof InstancePropertyReference;
        
        if (!$isCallExpr && !$isFieldRef)
            return null;
        // quick return
        
        $codeTmpl = null;
        $model = Array();
        $context = new VMContext(new ObjectValue($model), $this);
        
        $model["type"] = new TypeValue($expr->getType());
        foreach (array_keys($this->modelGlobals) as $name)
            $model[$name] = @$this->modelGlobals[$name] ?? null;
        
        if ($isCallExpr) {
            $call = $expr;
            $parentIntf = $call->getParentInterface();
            $methodName = ($parentIntf === null ? "" : $parentIntf->name . ".") . $call->getName() . "@" . count($call->args);
            $callTmpls = @$this->methods[$methodName] ?? null;
            if ($callTmpls === null)
                return null;
            
            foreach ($callTmpls as $callTmpl) {
                if ($expr instanceof InstanceMethodCallExpression)
                    $model["this"] = new ExpressionValue($expr->object);
                for ($i = 0; $i < count($callTmpl->args); $i++)
                    $model[$callTmpl->args[$i]] = new ExpressionValue($call->args[$i]);
                
                if ($callTmpl->template->ifExpr === null || ((new ExprVM($context))->evaluate($callTmpl->template->ifExpr))->value) {
                    $codeTmpl = $callTmpl->template;
                    break;
                }
            }
        }
        else if ($isFieldRef) {
            $cm = ($expr)->getVariable();
            $field = @$this->fields[$cm->getParentInterface()->name . "." . $cm->name] ?? null;
            if ($field === null)
                return null;
            
            if ($expr instanceof InstanceFieldReference || $expr instanceof InstancePropertyReference)
                $model["this"] = new ExpressionValue(($expr)->object);
            $codeTmpl = $field->template;
        }
        else
            return null;
        
        if ($codeTmpl === null)
            return null;
        
        foreach ($codeTmpl->includes ?? array() as $inc)
            $this->generator->addInclude($inc);
        
        $tmpl = (new TemplateParser($codeTmpl->template))->parse();
        $result = $tmpl->format($context);
        return $result;
    }
    
    function stmt($stmt) {
        return null;
    }
}
